Fix tox/linting

kfsone opened the pull request:
This addresses all the outstanding "flake8" issues and reduces the majority of pylint issues.

The remaining issues probably need looking at - some of them sound serious but a couple of them needed more time than I had to look at them.

A couple of the commits are a bit lengthy as they fix cosmetic errors in minor files, but I already caught one bug (a missing f infront of a format string)

<img width="2535" height="993" alt="image" src="https://github.com/user-attachments/assets/0434d43a-25db-4ec4-8eb0-d5c050d0ab44" />

kfsone committed:
chore: tidy up tox and improve usability

- removed linters from the default envlist,
- added python 3.14,
- removed older python versions,
- added python 3.14 to the github list,
- added 'skip_install' so that it doesn't always forcibly reinstall the requirements,
- added pylint to the dev-requirements

* chore because: does not require a release.

kfsone committed:
chore: updated code to comply with flake8 requirements

This means that you can now run "tox -e flake8" and get no errors.

kfsone committed:
fix: eddblink expiration wasn't correctly incremental

(caught by linter)

kfsone committed:
chore: linter issues and linter config so that flake8 is clean

this also makes the pylint output a little less overwhelming and more meaningful

kfsone committed:
chore: resolve numerous linter errors and making pylint near-viable

The remaining pylint issues mostly look like they deserve attention.

To run everything:
$ tox run

To run everything all at once (much faster, more chaotic output until you're down to just errors)
$ tox --parallel

To run a single test suite, e.g. py311 or flake8
$ tox -e flake8
$ tox -e flake8,pylint
$ tox -e flake8,pylint --parallel

Tromador commented:
Wow. How long did that lot take you? I'm just about to sleep. I'll look at it later though given that the log has massively diminished, I expect it's fine. Also, I am curious to see the diffs.

Tromador commented:
What's happened to lookup() in tradedb. A quick skim and I can see the vast majority is just stylistic choices to keep flake/lint happy, but it looks like you've removed that function completely and it's the only thing that really confuses me.

kfsone commented:
> What's happened to lookup() in tradedb. A quick skim and I can see the vast majority is just stylistic choices to keep flake/lint happy, but it looks like you've removed that function completely and it's the only thing that really confuses me.

Ahaha - oh, that's a beauty. I'm going to let you open it in an editor to see how the linter earned it's keep on that one.

kfsone commented:
<img width="655" height="400" alt="image" src="https://github.com/user-attachments/assets/aae4b709-7b1b-4182-b1bc-844fb9e1d5b6" />

see it?

kfsone commented:
The closer we get to consistent linting patterns, the more precise the linters can be about this kind of mistake.

TLDR: Somewhere, git got confused in a merge and kept both the before and after.

What we have is:

```
....def outer_method(self, ...):
....|...def _lookup(... new implementation, inline helper):
....|...|...stuff
....|...stuff still in outer_method
....|...raise YOU SHALL NOT PASS THIS
....|...def lookup(...):
....|...|... this is perfectly valid, it can just never be reached
```

The linter surfaced this with various warnings about weird variable uses, it wasn't drop-dead obvious "hey you have this over-indented function definition that's unreachable due to an unconditional raise",

It *could* be that precise (especially the 'ruff' linter, which I love), but we'd have to normalize closer to standard formatting.

kfsone commented:
Linters are a pain in the ass; who gives a crap about stupid style rules?

Well, my brain. Your brain. Their brains. When trying to look quickly/efficiently over a screenfull of code.

Especially in a language like Python where small things make a major difference:

```
1, 2, 3
1, 2
1,
0
```

what *type* is each of those lines? easy to spot the missing comma, right? but what if it's in an argument list with parens?
```
foo(that).bar(this, that).deploy(values=(1,2,3), prefixes=(1,2), primary_key=(1,), indexes=(0))
```

Linting is about making code easier to scan/read/maintain, with the bonus that once you are working *with* it, those stupid style things can become genuine useful code smell detectors.

It's complaining about the use of

```
Session = tdb.orm.Session
with Session.begin():
```

somewhere, because 'Session' is an import at the top level or something.

I tried changing it to

```
with tdb.orm.Session.begin():
```

and that broke. Which means, we're currently relying on scope leakage somewhere; something later is actually accessing globals()["Session"] and requiring that it got leakily set...

Tromador commented:
Sorry. I've been stuck in bed with my illness kicking my arse. I'll try and catch up with your comments and look at this today.

Tromador commented:
### tox.ini

skil_install, should be skip_install

You add a comment "# note: not all parts of this file support comments on data lines, so its best to put comments on separate lines. ftw."

Then add 
\+    consider-using-f-string,		# TODO: use fstrings :)
\+	broad-exception-caught,			# TODO: Be more surgical about catch

Which is exactly what you just said not to do.

### eddlink_plug.py

```
assert headers[:10] == expect_headers, \
    "unrecognized listings csv format. expected {expected_headers}; got {headers}"
```

Shouldn't this be an f-string? And is this the correct variable name, or should it be expect_headers. Maybe also better as a proper exception not an assert?

```
def finish(self):
    raise RuntimeError("unexpected call to finish")
```

This might come down to a style choice. Because run() correctly returns false, we don't need finish() at all. If you're adding it to be complete, or because lint reasons, maybe just return false and perhaps add a comment if you want saying we should never get here? To me (and this is just my opinion) adding explosions to satisfy completeness, or lint, is too much.

### couple of places

raise e from e

Doesn't this make the exception its own cause and so will screw with tracebacks? "This exception was caused by itself" isn't normally a useful error message. If anything it's just noise. If you meant to “propagate unchanged”, maybe use raise (bare) or remove the except. If the intention is “wrap with context”, use raise SomeError(...) from e.

### the rest

Other than those small things, I really like this. I'm totally on the same page about making shit run through a lint checker and that being potentially helpful in finding stupid errors. Not sure how @eyeonus feels about it, but I don't imagine he would object, so long as he isn't volunteered for the work!

kfsone commented:
> ### tox.ini
> skil_install, should be skip_install
> 
> You add a comment "# note: not all parts of this file support comments on data lines, so its best to put comments on separate lines. ftw."
> 
> Then add + consider-using-f-string, # TODO: use fstrings :) + broad-exception-caught, # TODO: Be more surgical about catch
> 
> Which is exactly what you just said not to do.

It's demonstrably already working there :-P I'll change the wording of the caution subtly.
 
> ### eddlink_plug.py
> ```
> assert headers[:10] == expect_headers, \
>     "unrecognized listings csv format. expected {expected_headers}; got {headers}"
> ```
> 
> Shouldn't this be an f-string? And is this the correct variable name, or should it be expect_headers. Maybe also better as a proper exception not an assert?

Ahh, yes and yes. 

> ```
> def finish(self):
>     raise RuntimeError("unexpected call to finish")
> ```
> 
> This might come down to a style choice. Because run() correctly returns false, we don't need finish() at all. If you're adding it to be complete, or because lint reasons, maybe just return false and perhaps add a comment if you want saying we should never get here? To me (and this is just my opinion) adding explosions to satisfy completeness, or lint, is too much.

That's a good call. It's not really about satisfying the linter, it's rather about squaring the tool away and making it so a future maintainer knows we didn't just forget something. Framing: that's the term?

a- could add a linter hint about it, but the linter is just trying to be helpful and my design for the base was intentional here, so it's following thru for me,
b- could repeat the base class, but then a future maintainer can't tell if I was c&p lazy or I have to annotate why I think it should be a fatal exception if someone does call finish on us (would be fair: it's a contract; run returns X, so I'm expecting you NOT to call my finish),
c- could return False, but if that happens then how is anyone supposed to know that we were *relying* on it not being called?
d- current implementation, except I was just being lazy tbh; I meant to run it and see if it got triggered. it didn't, and I forgot :(

> ### couple of places
> raise e from e
> 
> Doesn't this make the exception its own cause and so will screw with tracebacks? "This exception was caused by itself" isn't normally a useful error message. If anything it's just noise. If you meant to “propagate unchanged”, maybe use raise (bare) or remove the except. If the intention is “wrap with context”, use raise SomeError(...) from e.

They need qualifying, I think - will check. My limited experience was that "e from e" is a way to indicate you're handling other exceptions and specifically want to call out that it's bad to get this particular exception in this scenario.

```
try:
  sip_coffee()
except EmptyCup:
  refill()
except BleachInCup as e:
  raise CaffeineError("I WAS GOING TO DRINK THAT, MEG") from e
```

I loathe Python tracebacks in the first place, and I use "from <something>" other than "None" with great reluctance, but they absolutey _do not_ say "with great reluctance comes great correctness" and I'll revisit these.

That is: sometimes I have to conceed that not knowing what the underlying exception was will be annoying, but just letting it fall through without decoration might be equally frustrating. On the *very* rare occasion, I find myself thinking "I might actually regret not having ALL of the shite this is going to vomit slightly more than I'll regret having it vomitted all of the shite at me :)"

kfsone committed:
address review feedback and complete pass.

chore: reduce pylint errors,
feat: add support for ruff (not actively enabled)
chore: address emerging flake8 warnings,
chore: disable noisy flake8/pylint warnings but with annotations,
chore: disable ruff warnings that conflict with repos code styles,
chore: annotate noisy ruff warnings that need an opine,
chore: address pylint/ruff warnings,
chore: type hinting
chore: tidy up how certain plugins/commands communicate with caller (False -> you don't need to do anything for me)
"

kfsone commented:
incoming delta with fixes above and a finishing pass thru things to get the warnings out of pylint down to a reasonable level.

New file: .ruff.toml: 

Disclosure: I'm a increasingly using Astral's python tooling, so added a ".ruff.toml" here. Lets *me* bounce things off `ruff` (sales pitch deferred). TLDR; it's 5-30x faster than flake8 and does a hell of a lot more.

```
. ./.venv/scripts/activate.ps1 ; tox -e flake8,ruff --parallel  # flake8 will win the first time while ruff installs.
```

I'm also using 'uv' instead of pip/poetry/conda/pyenv/all-the-other-goddamn-package-managers

(from a presentation I gave: 
<img width="400" height="225" alt="Snakes0022" src="https://github.com/user-attachments/assets/36ddaccb-fb6a-4d4b-8ece-27fb699ae538" />)

For simplicity: uv = compiled pip, not a python dependency... can install the-python-you-need to do everything without first requiring you to get past having a _suitable_ python to run whatever bootstrap you have.

uv is also 10-100x faster than pip in my experience, even if it has to download a python to install into a .venv...

So I *also* have some accompanying "pyproject.toml" changes, but I'm holding those back so as not to inflict workflow changes.

However: uv is really good for building/deploying/distributing packages; infact the only PyPI packages I've ever published I did with uv because I considered the effort required to make it do it acceptable :)

kfsone committed:
chore: comma separate envs?

Tromador commented:
What is this function and how does it relate to the lint pass? As for all the uv stuff, I'm going to defer to @eyeonus and see what he thinks, as you're getting into publishing stuff and I don't think I should be making those calls for him.

``````
def update_database_schema(tdb: TradeDB | TradeORM) -> None:
    """ Check if there are database changes to be made, and if so, execute them. """
    # TODO: This should really be a function of the DB itself and not something
    # the caller has to ask the database to do for it.
    template_folder = getattr(tdb, "templatePath", None)
    if not template_folder:
        return

    db_change = Path(template_folder, "database_changes.json")
    if not db_change.exists():
        return

    try:
        with db_change.open("r", encoding="utf-8") as file:
            for change in ijson.items(file, 'item'):
                tdb.getDB().execute(change)
    finally:
        db_change.unlink()
```````

kfsone commented:
> What is this function and how does it relate to the lint pass? As for all the uv stuff, I'm going to defer to @eyeonus and see what he thinks, as you're getting into publishing stuff and I don't think I should be making those calls for him.

I'm not fishing to push for that, I was explaining why I was "sneaking" in a file that is probably only relevant to me, which I think otherwise would look passive-aggressive?

> def update_database_schema(tdb: TradeDB | TradeORM) -> None:
>     """ Check if there are database changes to be made, and if so, execute them. """
>     # TODO: This should really be a function of the DB itself and not something
>     # the caller has to ask the database to do for it.
>     template_folder = getattr(tdb, "templatePath", None)
...
>     finally:
>         db_change.unlink()

This was existing code I was just isolating it into its own function so run() was less cluttered:

<img width="704" height="219" alt="image" src="https://github.com/user-attachments/assets/5ca323b9-d985-44e5-895c-5750607315d2" />

Tromador commented:
Oliver mate - you shoulda just said, 
"I'm adding this file because I need it to make 'ruff' work. If you're not familiar with it, here's a link. It does a few different things, but I'm used to using it for linting and having that file will make my life easier.

If you had said that, I wouldn't even have questioned it, but the way you presented it as best thing since sliced bread including the confusing picture from your presentation and how wonderful it is and all the things it can do and replace pip and publish stuff and terribly enthused about the whole thing and you were looking like a second hand 'Astral' salesman on a commission only salary.

I keep telling people I'm a simple man, nobody ever believes me. Give me simple information. Had you randomly added it, I'd probably have asked "woss dis den?" but I wouldn't find it passive-aggressive.

I wonder how far back that db_change code goes? It's possible I am oversensitive to stuff that has the letter b following d - especially in tradedb which as discussed makes its own internal db, from the db and my brain starts to implode at that point.

Tromador commented:
Ok, the only thing then I'm still concerned about is the 'raise e from e' stuff. You said you'd revisit them - did you decide they are correct, if so please can you explain, or did you miss some?

One in cache.py, two in tradedb.py

eyeonus commented:
> "# note: not all parts of this file support comments on data lines, so its best to put comments on separate lines. ftw."

I just want to put my two cents in and say, while not like, an official formatting requirement or anything like that, I do prefer comments to be on the preceding line.

[Still catching up on this, will edit if I have anything to add after reading more]

> Shouldn't this be an f-string? And is this the correct variable name, or should it be expect_headers. Maybe also better as a proper exception not an assert?

We are incredibly inconsistent with f-strings, mainly because a lot of this was written before that was a thing, which is why eg. `tdenv.NOTE()` and family exist:

```
    tdenv.DEBUG1("hello, {world}{}", "!", world="world")

is equivalent to:

    if tdenv.debug >= 1:
        print("#hello, {world}{}".format("!", world="world"))
```

I am perfectly fine with having linting be a thing, I like type hinting and all that. Modernizing the code to take advantage of all the various pythonic improvements is a great idea and I'm all for it. I would even do it myself if I had the energy and the time in sufficient quantity. Working on this is supposed to be part of my practical education in programming while I'm doing the whole school thing. Man do I wish I lived in a country that paid for people to do college.


> That's a good call. It's not really about satisfying the linter, it's rather about squaring the tool away and making it so a future maintainer knows we didn't just forget something. Framing: that's the term?

May I suggest:
```
def finish(self):
    """
        Is not used.
    """
    pass
```

[This space intentionally left blank.]

eyeonus commented:
> What is this function and how does it relate to the lint pass? As for all the uv stuff, I'm going to defer to @eyeonus and see what he thinks, as you're getting into publishing stuff and I don't think I should be making those calls for him.

I mean it sounds interesting. The advantages sound suitably advantageous. I know how to use pyenv, barely, so I'm not committed to any particular venv method. If uv makes things easier then by all means.

We use python-semantic-release to publish because the whole automatic version incrementing thing is just too great to not, so unless uv can replicate what that does I wouldn't consider switching the publishing pathway, however.

kfsone commented:
> > Shouldn't this be an f-string? And is this the correct variable name, or should it be expect_headers. Maybe also better as a proper exception not an assert?
> 
> We are incredibly inconsistent with f-strings, mainly because a lot of this was written before that was a thing, which is why eg. `tdenv.NOTE()` and family exist:

Ah - he was correctly spotting a pair of mistakes on my part :)
> 
> ```
>     tdenv.DEBUG1("hello, {world}{}", "!", world="world")
> 
> is equivalent to:
> 
>     if tdenv.debug >= 1:
>         print("#hello, {world}{}".format("!", world="world"))
> ```

That's a fossil of Python evolution. ".format" is a function call that requires a method-lookup on the string object being formatted against.

> I am perfectly fine with having linting be a thing, I like type hinting and all that. Modernizing the code to take advantage of all the various pythonic improvements is a great idea and I'm all for it. I would even do it myself if I had the energy and the time in sufficient quantity. Working on this is supposed to be part of my practical education in programming while I'm doing the whole school thing. Man do I wish I lived in a country that paid for people to do college.

There are two schools of thought re linting/checking/testing/yadda.

Those who say it is holy divine magic that Makes All Right.

And those who say it's make-work bull-shit distracting you from the real engineering work imposed by shitty coders who are infuriated that HR won't let them install webcams inside your keyboard to monitor your every keypress.

I think the answer is that these are tools and if you use them in certain ways they take an actual load off you.

It's a linters' job to act all butt-hurt opinionated, it's "our" job to remember their opinion is whatever values we type into their config file, and until we do, probably some default values. And it's not, like, the program's *actual* opinion :)

> May I suggest:
> 
> ```
> def finish(self):
>     """
>         Is not used.
>     """
>     pass
> ```

Bear in mind that Python functions *always* actually return a value, so in the above you're _implicitly_ returning None, and way back the calling code was stupid and cared about the difference between True, False, and None...

The purpose of a NotImplementedError is to tell downstream developers that there's a contract, and finish() is expected to return True or False; True: please do more things, False: stop now, we're done.

At the same time, the contract is also supposed to be that if run() returns False, finish() doesn't get called.

It really doesn't matter one way or the other, because finish() never gets called, but I've been on the hook on call for the side-effects of a can-kick like this to be easy with some of the solutions myself :)

kfsone commented:
> > What is this function and how does it relate to the lint pass? As for all the uv stuff, I'm going to defer to @eyeonus and see what he thinks, as you're getting into publishing stuff and I don't think I should be making those calls for him.
> 
> I mean it sounds interesting. The advantages sound suitably advantageous. I know how to use pyenv, barely, so I'm not committed to any particular venv method. If uv makes things easier then by all means.
> 
> We use python-semantic-release to publish because the whole automatic version incrementing thing is just too great to not, so unless uv can replicate what that does I wouldn't consider switching the publishing pathway, however.

Oh, I'm absolutely not suggesting or hinting. I'm crazy stoked about it for unrelated reasons and stopping there :)


@tromador
tradedangerous/cache.py:
<img width="985" height="198" alt="image" src="https://github.com/user-attachments/assets/265ad7c9-b5b8-405f-8de3-1340056ef23b" />
The original intent was "actually, under this circumstance, *don't* catch the exception". So, reduced to just "raise".

tradedb.py in lookupPlace:

<img width="769" height="276" alt="image" src="https://github.com/user-attachments/assets/836118cd-028b-48ea-95be-ec71557e44a8" />
Looks like the same idea, but I think we actually want to just catch the lookup error.

<img width="676" height="303" alt="image" src="https://github.com/user-attachments/assets/21999a09-9100-47e2-900a-a32dede676a5" />
I'm going to leave this one as is; usually if you see an exception, it means everything stopped at the callsite. You don't expect an intermediate exception handler to have possibly done other stuff?

<img width="687" height="295" alt="image" src="https://github.com/user-attachments/assets/2970cb0c-6292-49a6-88ec-91f715cdad4b" />
I'm going to leave this one; since we're explicitly changing the exception type and explicitly to one that tries to clean-up the error output, you're not going to *see* the double backtrace unless you're running it in a debugger or something and then you probably want to know where this exception came from...

kfsone committed:
chore: review feedback

Tromador commented:
On the remaining raise e from e in writePersist: I get the intent (you’re basically saying “if we’re here, something has gone wrong and we want it to be loud”), but I still think raise e from e is a slightly odd construct because it makes the exception its own __cause__ and tends to add traceback noise without adding real information. 

If the goal is “propagate unchanged as if we didn’t catch it”, a bare raise like you did in cache.py keeps the original traceback intact.

If you want to “add context because this specific PermissionError is special”, then raise TradeException(msg) from e (or from None if you to suppress the OS-level cause). 

That said, I appreciate that you've listened to everything I've said and we can agree to disagree over one line which isn't a showstopper bug, so I'll go ahead and merge this.

kfsone commented:
> If you want to “add context because this specific PermissionError is special”, then raise TradeException(msg) from e (or from None if you to suppress the OS-level cause).

That nuance totally went over my head on this one - you're 100% right there. tbds
